### PR TITLE
Emit warning when assigned string value is truncated

### PR DESCRIPTION
**Pull request by taldcroft**
This is a partial fix for #5551 so that a warning is emitted if a string value gets truncated when assigning in-place to a string column item.

There will be a performance penalty, but I think it is not too bad given that setitem is already using pure-Python code.  Quick timeit testing showed that the difference is within the run-to-run uncertainty.

### CODE REVIEW

File: astropy/table/tests/test_column.py [hunk 1]
```diff
@@ -547,3 +547,39 @@ def test_qtable_column_conversion():
...
+
+def test_string_truncation_warning():
+    """
+    Test warnings associated with in-place assignment to a string
+    column that results in truncation of the right hand side.
+    """
+    t = table.Table([['aa', 'bb']], names=['a'])
+
+    with catch_warnings() as w:
+        from inspect import currentframe, getframeinfo
```
Comment (hamogu) [inline astropy/table/tests/test_column.py]: Not needed here. Move to the next `catch_warnings` block for readability

Comment (taldcroft) [inline astropy/table/tests/test_column.py]: OK, I have done this locally.

**Comment by astrofrog** [comment]
Great! This has been bugging me for a while 🎉

**Comment by taldcroft** [comment]
#5700 has been bugging me for 4 years...

**Pull request merged**